#!/usr/bin/env python3

import rospy
import numpy as np
import cv2
import threading
import tf2_ros
import onnxruntime
from lanelet2.io import Origin, load
from lanelet2.projection import UtmProjector
from image_geometry import PinholeCameraModel
from shapely.geometry import LineString

from geometry_msgs.msg import Point, PointStamped
from sensor_msgs.msg import Image
from sensor_msgs.msg import CameraInfo
from autoware_mini.msg import TrafficLightResult, TrafficLightResultArray
from autoware_mini.msg import Path
from tf2_geometry_msgs import do_transform_point

from cv_bridge import CvBridge, CvBridgeError

CLASSIFIER_RESULT_TO_STRING = {
    0: "green",
    1: "yellow",
    2: "red",
    3: "unknown"
}

CLASSIFIER_RESULT_TO_COLOR = {
    0: (0, 255, 0),
    1: (255, 255, 0),
    2: (255, 0, 0),
    3: (0, 0, 0)
}

CLASSIFIER_RESULT_TO_TLRESULT = {
    0: 1,  
    1: 0,  
    2: 0,  
    3: 2  
}


class CameraTrafficLightDetector:
    def __init__(self):

        
        onnx_path = rospy.get_param("~onnx_path")
        self.rectify_image = rospy.get_param('~rectify_image')
        self.roi_width_extent = rospy.get_param("~roi_width_extent")
        self.roi_height_extent = rospy.get_param("~roi_height_extent")
        self.min_roi_width = rospy.get_param("~min_roi_width")
        self.transform_timeout = rospy.get_param("~transform_timeout")
        
        coordinate_transformer = rospy.get_param("/localization/coordinate_transformer")
        use_custom_origin = rospy.get_param("/localization/use_custom_origin")
        utm_origin_lat = rospy.get_param("/localization/utm_origin_lat")
        utm_origin_lon = rospy.get_param("/localization/utm_origin_lon")
        lanelet2_map_path = rospy.get_param("~lanelet2_map_path")

        self.trafficlights = None
        self.tfl_stoplines = None
        self.camera_model = None
        self.stoplines_on_path = None

        self.lock = threading.Lock()
        self.bridge = CvBridge()
        self.model = onnxruntime.InferenceSession(onnx_path, providers=['CUDAExecutionProvider'])
        self.tf_buffer = tf2_ros.Buffer()
        self.tf_listener = tf2_ros.TransformListener(self.tf_buffer)

        
        if coordinate_transformer == "utm":
            projector = UtmProjector(Origin(utm_origin_lat, utm_origin_lon), use_custom_origin, False)
        else:
            raise RuntimeError('Only "utm" is supported for lanelet2 map loading')
        lanelet2_map = load(lanelet2_map_path, projector)

        
        all_stoplines = get_stoplines(lanelet2_map)
        self.trafficlights = get_stoplines_trafficlights(lanelet2_map)
        
        self.tfl_stoplines = {k: v for k, v in all_stoplines.items() if k in self.trafficlights}

        
        self.tfl_status_pub = rospy.Publisher('traffic_light_status', TrafficLightResultArray, queue_size=1,
                                              tcp_nodelay=True)
        self.tfl_roi_pub = rospy.Publisher('traffic_light_roi', Image, queue_size=1, tcp_nodelay=True)

        
        rospy.Subscriber('camera_info', CameraInfo, self.camera_info_callback, queue_size=1, tcp_nodelay=True)
        rospy.Subscriber('/planning/local_path', Path, self.local_path_callback, queue_size=1, buff_size=2 ** 20,
                         tcp_nodelay=True)
        rospy.Subscriber('image_raw', Image, self.camera_image_callback, queue_size=1, buff_size=2 ** 26,
                         tcp_nodelay=True)

    def camera_info_callback(self, camera_info_msg):
        if self.camera_model is None:
            camera_model = PinholeCameraModel()
            camera_model.fromCameraInfo(camera_info_msg)
            self.camera_model = camera_model

    def local_path_callback(self, local_path_msg):

        if len(local_path_msg.waypoints) == 0:
            with self.lock:
                self.stoplines_on_path = []
                self.transform_from_frame = local_path_msg.header.frame_id
            return
        
        path_points = []
        for pose in local_path_msg.waypoints:
            path_points.append((pose.position.x, pose.position.y))
        
        local_path_line = LineString(path_points)

        stoplines_on_path = []
        
        for stopline_id, stopline_geom in self.tfl_stoplines.items():
            if local_path_line.intersects(stopline_geom):
                stoplines_on_path.append(stopline_id)
        
        with self.lock:
            self.stoplines_on_path = stoplines_on_path
            self.transform_from_frame = local_path_msg.header.frame_id

        
    def camera_image_callback(self, camera_image_msg):

        if self.camera_model is None:
            rospy.logwarn_throttle(10, "%s - No camera model received, skipping image", rospy.get_name())
            return

        if self.stoplines_on_path is None:
            rospy.logwarn_throttle(10, "%s - No path received, skipping image", rospy.get_name())
            return

        with self.lock:
            stoplines_on_path = self.stoplines_on_path
            transform_from_frame = self.transform_from_frame
        
        if len(self.stoplines_on_path) == 0:
            return
        
        transform = self.tf_buffer.lookup_transform(camera_image_msg.header.frame_id, 
                                                        transform_from_frame,
                                                        camera_image_msg.header.stamp,
                                                        rospy.Duration(self.transform_timeout))
            

        image = self.bridge.imgmsg_to_cv2(camera_image_msg,  desired_encoding='rgb8')
        
        if self.rectify_image:
            self.camera_model.rectifyImage(image, image)
                
        rois = self.calculate_roi_coordinates(stoplines_on_path=stoplines_on_path, transform=transform)
        
        tfl_status = TrafficLightResultArray()
        tfl_status.header.stamp = camera_image_msg.header.stamp
        
        if len(rois) != 0:
            roi_images = self.create_roi_images(image, rois=rois)
            predictions = self.model.run(None, {'conv2d_1_input': roi_images})[0]

            classes, scores = [], []
            for each_pred in predictions:
                classes.append(np.argmax(each_pred))
                scores.append(np.max(each_pred))

            for cl, (stoplineId, plId, _, _, _, _) in zip(classes, rois):
                            
                tfl_result = TrafficLightResult()
                tfl_result.light_id = plId
                tfl_result.stopline_id = stoplineId
                tfl_result.recognition_result =  CLASSIFIER_RESULT_TO_TLRESULT[cl]
                tfl_result.recognition_result_str = CLASSIFIER_RESULT_TO_STRING[cl]

                tfl_status.results.append(tfl_result)
                
        self.tfl_status_pub.publish(tfl_status)


        self.publish_roi_images(image, rois, classes, scores, camera_image_msg.header.stamp)


    def calculate_roi_coordinates(self, stoplines_on_path, transform):
        
        rois = []

        for linkId in stoplines_on_path:
            for plId, traffic_lights in self.trafficlights[linkId].items():
                us = []
                vs = []

                for x, y, z in traffic_lights.values():
                    point_map = Point(float(x), float(y), float(z))
                    point_camera = do_transform_point(PointStamped(point=point_map), transform).point
                    u,v = self.camera_model.project3dToPixel((point_camera.x, point_camera.y, point_camera.z))
                    if u < 0 or u >= self.camera_model.width or v < 0 or v >= self.camera_model.height:
                        continue

                    extent_x_px = self.camera_model.fx() * self.roi_width_extent / point_camera.z
                    extent_y_px = self.camera_model.fy() * self.roi_height_extent / point_camera.z

                    us.extend([u + extent_x_px, u - extent_x_px])
                    vs.extend([v + extent_y_px, v - extent_y_px])

                if len(us) < 8:
                    continue

                us = np.clip(np.round(np.array(us)), 0, self.camera_model.width - 1)
                vs = np.clip(np.round(np.array(vs)), 0, self.camera_model.height - 1)

                min_u = int(np.min(us))
                max_u = int(np.max(us))
                min_v = int(np.min(vs))
                max_v = int(np.max(vs))

                if max_u - min_u < self.min_roi_width:
                    continue

                rois.append([int(linkId), plId, min_u, max_u, min_v, max_v])

        return rois


    def create_roi_images(self, image, rois):
        roi_images = []
        for _, _, min_u, max_u, min_v, max_v in rois:
            selected_area = image[min_v:max_v, min_u:max_u]
            resized_image = cv2.resize(selected_area, (128, 128), interpolation=cv2.INTER_LINEAR)
            resized_image = resized_image.astype(np.float32)
            roi_images.append(resized_image)

        return np.stack(roi_images, axis=0) / 255.0


    def publish_roi_images(self, image, rois, classes, scores, image_time_stamp):

        if len(rois) > 0:
            for cl, score, (_, _, min_u, max_u, min_v, max_v) in zip(classes, scores, rois):
                text_string = "%s %.2f" % (CLASSIFIER_RESULT_TO_STRING[cl], score)
                text_width, text_height = cv2.getTextSize(text_string, cv2.FONT_HERSHEY_SIMPLEX, 1.5, 2)[0]
                text_orig_u = int(min_u + (max_u - min_u) / 2 - text_width / 2)
                text_orig_v = max_v + text_height + 3

                start_point = (min_u, min_v)
                end_point = (max_u, max_v)
                cv2.rectangle(image, start_point, end_point, color=CLASSIFIER_RESULT_TO_COLOR[cl], thickness=3)
                cv2.putText(image,
                            text_string,
                            org=(text_orig_u, text_orig_v),
                            fontFace=cv2.FONT_HERSHEY_SIMPLEX,
                            fontScale=1.5,
                            color=CLASSIFIER_RESULT_TO_COLOR[cl],
                            thickness=2)

        image = cv2.resize(image, None, fx=0.5, fy=0.5, interpolation=cv2.INTER_LINEAR)
        img_msg = self.bridge.cv2_to_imgmsg(image, encoding='rgb8')

        img_msg.header.stamp = image_time_stamp
        self.tfl_roi_pub.publish(img_msg)

    def run(self):
        rospy.spin()


def get_stoplines(lanelet2_map):
    stoplines = {}
    for line in lanelet2_map.lineStringLayer:
        if line.attributes:
            if line.attributes["type"] == "stop_line":
                stoplines[line.id] = LineString([(p.x, p.y) for p in line])

    return stoplines


def get_stoplines_trafficlights(lanelet2_map):

    signals = {}

    for reg_el in lanelet2_map.regulatoryElementLayer:
        if reg_el.attributes["subtype"] == "traffic_light":
            linkId = reg_el.parameters["ref_line"][0].id

            for tfl in reg_el.parameters["refers"]:
                tfl_height = float(tfl.attributes["height"])
                plId = tfl.id

                traffic_light_data = {'top_left': [tfl[0].x, tfl[0].y, tfl[0].z + tfl_height],
                                      'top_right': [tfl[1].x, tfl[1].y, tfl[1].z + tfl_height],
                                      'bottom_left': [tfl[0].x, tfl[0].y, tfl[0].z],
                                      'bottom_right': [tfl[1].x, tfl[1].y, tfl[1].z]}

                signals.setdefault(linkId, {}).setdefault(plId, traffic_light_data)

    return signals

if __name__ == '__main__':
    rospy.init_node('camera_traffic_light_detector', log_level=rospy.INFO)
    node = CameraTrafficLightDetector()
    node.run()